src/main/java/uk/ac/ebi/ampt2d/accession/common/accessioning/BasicAccessioningService.java

```diff
@@ -0,0 +1,136 @@
+                                    DigestFunction<MODEL> digestFunction,
+                                    HashingFunction<HASH> hashingFunction) {
+        this.accessionGenerator = accessionGenerator;
+        this.dbService = dbService;
+        this.digestFunction = digestFunction;
+        this.hashingFunction = hashingFunction;
+    }
+
+    /**
+     * Get accessions for a list of messages. It looks for the object's accessions in a repository, and it they don't
+     * exist, generate new ones, storing them in the repository
+     *
+     * @param messages
+     * @return
+     */
+    @Override
+    public Map<ACCESSION, MODEL> getOrCreateAccessions(List<? extends MODEL> messages) {
+        Map<HASH, MODEL> hashToMessages = mapHashOfMessages(messages);
+        Map<HASH, ACCESSION> existingAccessions = dbService.getExistingAccessions(hashToMessages.keySet());
+        Map<HASH, MODEL> newMessages = filterNotExistingAccessions(hashToMessages, existingAccessions);
+
+        Map<ACCESSION, MODEL> accessions = joinExistingAccessionsWithMessages(existingAccessions, hashToMessages);
+        if (!newMessages.isEmpty()) {
+            accessions.putAll(generateAccessions(newMessages));
+        }
+        return accessions;
+    }
+
+    /**
+     * Digest messages, hash them and map them. If Two messages have the same hash keep the first one.
+     *
+     * @param messages
+     * @return
+     */
+    private Map<HASH, MODEL> mapHashOfMessages(List<? extends MODEL> messages) {
+        return messages.stream().collect(Collectors.toMap(digestFunction.andThen(hashingFunction), e -> e, (r, o) -> r));
+    }
+
+    private Map<HASH, MODEL> filterNotExistingAccessions(Map<HASH, MODEL> hashToMessages,
+                                                         Map<HASH, ACCESSION> existingAccessions) {
+        return hashToMessages.entrySet().stream()
+                .filter(entry -> !existingAccessions.containsKey(entry.getKey()))
+                .collect(Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue()));
+    }
+
+    private Map<ACCESSION, MODEL> joinExistingAccessionsWithMessages(Map<HASH, ACCESSION> existingAccessions,
+                                                                     Map<HASH, MODEL> hashToMessages) {
+        return existingAccessions.values().stream().collect(Collectors.toMap(e -> e, e -> hashToMessages.get(e)));
+    }
+
+    private Map<ACCESSION, MODEL> generateAccessions(Map<HASH, MODEL> accessions) {
+        SaveResponse<ACCESSION, MODEL> response = dbService.save(accessionGenerator.generateAccessions(accessions));
+        accessionGenerator.postSave(response);
+        return response.getAllAccessionToMessage();
+    }
+
+    @Override
+    public Map<ACCESSION, MODEL> getAccessions(List<? extends MODEL> accessionObjects) {
+        return dbService.findAllAccessionByMessageHash(getHashes(accessionObjects));
```

Comment:
This method is never used.

Reply:
We have it as a use case, had this discussion yesterday offline.

Reply:
so ok we are gonna have two endpoints one for getorcreate and one for getaccession ?

Reply:
Not for now.